router/pkg/watcher/watcher_test.go

```diff
@@ -806,9 +898,149 @@ func TestCancel(t *testing.T) {
 	require.ErrorIs(t, eg.Wait(), context.Canceled)
 }
 
-// sendTick helper function which adds a sleep timeout
-// so users don't need to manually add sleep to every test
-func sendTick(channel chan time.Time) {
+// sendSyncTick is a helper function for tests that are run
+// inside a synctest bubble. It sends a tick to the channel and waits
+// for all other goroutines apart from the current time advance inside of the current bubble to be blocked.
+func sendSyncTick(channel chan time.Time) {
 	channel <- time.Now()
 	time.Sleep(10 * time.Millisecond)
+	synctest.Wait()
+}
+
+// newFakeFileInfo is a helper function for tests that creates a new fake file info object
+// newFakeFileInfo creates a fakeFileInfo with modTime initialized to time.Now().
+// In synctest, time.Now() returns the fake time controlled by the test.
+func newFakeFileInfo() *fakeFileInfo {
+	return &fakeFileInfo{
+		modTime: time.Now(),
+	}
+}
+
+// fakeFileInfo wraps os.FileInfo to manually track file modification times in tests.
+//
+// Why we need this when using synctest:
+//
+// When you use synctest, time doesn't advance on its own - you control it. But here's the problem:
+// when you call os.WriteFile() or os.Rename(), the filesystem sets the file's mtime to the *real*
+// system time, not the fake time that synctest controls. This creates a mismatch.
+//
+// Even worse, many filesystems (like ext4 on Linux) don't support nanosecond precision for mtimes.
+// They might only do seconds or milliseconds. So if you write/rename files quickly in succession,
+// they can end up with identical timestamps from the filesystem's perspective.
+//
+// The watcher detects changes by calling os.Stat() and checking if stat.ModTime().After(prevModTime).
+// If timestamps don't advance (filesystem resolution issue) or don't match the test's time control
+// (synctest mismatch), this comparison breaks and the watcher won't detect changes properly.
+//
+// This wrapper fixes it by intercepting file operations and manually bumping modTime using time.Now()
+// (which respects synctest's fake time). Then we return our tracked modTime instead of whatever
+// the filesystem reports.
+type fakeFileInfo struct {
+	os.FileInfo
+	modTime time.Time
+}
+
+type fakeFS struct {
+	fileInfos map[string]*fakeFileInfo
+}
```

Comment:
maybe there's a better name than `fakeFS`? It's not really _fake_, just hacked to be more precise?